### DISCUSSION TITLE
Plugin Guide

### DISCUSSION THREAD

**Discussion by hanford** [TanStack/table#2471] [General]
Thanks, @tannerlinsley for all of your work in the react ecosystem.

React Table is one of my favorite go-to libraries at the moment. It's helped me build several complex yet performant UI's.

I was wondering if there was a Plugin Guide somewhere... I recently was able to modify `useRowSelect` to fit my needs, and once I kind of slowed down and truly read and understood the code I felt like I "got" plugins.. But it seems like this area of the project could use a little more documentation

Anyways, thanks again for the library!

**Comment by wbolduc** [comment]
Actually that plugin guide is alluded to at the end of [this section](https://github.com/tannerlinsley/react-table/blob/master/docs/api/README.md#the-stages-of-react-table-and-plugins) but it’s not clear where that actually is and there is no link

“To dive deeper into plugins, see Plugins and the Plugin Guide”

Bump to see this issue resolved, I’m very interested in seeing this document

**Reply by aslansservant42** [reply]
I too am looking for the plugin guide mentioned in that section of react-table doc. 

I have an idea for a horrizontal scroll plugin that could let buttons scroll left and right one column at a time by hiding and showing the non primary columns. That way the key/control column (usually column 0) stays and the user can scroll the information left and right.

**Comment by wbolduc** [comment]
Old reply, I have some notes from reading the code. They're not all the pretty but it's a start

useOptions; //(default_options) => (new options)

stateReducers; // (state, action, state_before_reduce, instance) => (new_state),
useControlledState; // (state_after_state_reducers, {instance}) => (new_state),

//decorating or modifying the user's columns
columns; //(columns, {instance}) => (new_columns)
columnsDeps; //(deps, {instance}) => depenencies for the memoized columns

//decorating or modifying all the columns (columns could have been added in the above step)
allColumns; //(columns, {instance}) => (new_columns)
allColumnsDeps; //(deps, {instance}) => dependencies for the memoized allColumns

// Allow plugins to manipulate the column value
accessValue; //(column_id, {row, column, instance}) => ??altered row value?

useInstanceAfterData; //(instance) => mutations to instance

//decorating the columns //default column is applied at the end of this
visibleColumns; //(columns, {instance}) => (new_columns)
visibleColumnsDeps; //(deps, {instance}) => dependencies to the memoized visibleColumns

headerGroups; //(header_group_data, instance) => ????new_header_group??
headerGroupsDeps; //(deps, {instance}) => dependencies to the memoized headerGroups

useInstanceBeforeDimensions; //(instance) => instance mutations
useInstance; //(instance) => instance mutations
getHeaderProps; //(prevProps, {instance, column, userProps}) => dom props
getFooterProps; //(prevProps, {instance, column, userProps}) => dom props
getHeaderGroupProps; //(prevProps, {instance, headerGroup, index, userProps}) => dom props
getFooterGroupProps; //(prevProps, {instance, headerGroup, index, userProps}) => dom props
getRowProps; //(prevProps, {instance, row, userProps}) => dom props
getCellProps; //(prevProps, {instance, cell, userProps}) => dom props

// need to apply any row specific hooks (useExpanded requires this)
prepareRow; // (row, {instance}) => row mutations

getTableProps; // (prevProps, {instance, userProps}) => dom props
getTableBodyProps; // (prevProps, {instance, userProps}) => dom props

useFinalInstance; //(instance) => mutations to instance

**Reply by aslansservant42** [reply]
thanks for posting this. How do we use this information? How do I modify the headerGroups before useTable() returns?
I have tried using useControlledState() and stateReducer(). With both the "state" object only consists of :
``` state;
{columnOrder: Array(0), expanded: {…}, hiddenColumns: Array(0)}
columnOrder: Array(0)
length: 0
[[Prototype]]: Array(0)
expanded:
53336: true
[[Prototype]]: Object
hiddenColumns: []
[[Prototype]]: Object
```

I was expecting 'state' to include columns, hiddenColumns, headerGroups, etc.
Also, with the reducer, the only "action" that I get are two calls for "init".

**Reply by tannerlinsley** [reply]
BTW, plugins are being deprecated. I would suggest moving to v8 and rearchitecting away from plugins to a more composable pattern.
